File: demos/vamp/MotionBenchmakerDemo.cpp
#include "MotionBenchmakerDemo.h"
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/json_parser.hpp>
#include <ompl/base/spaces/RealVectorStateSpace.h>
#include <ompl/base/ProblemDefinition.h>
#include <ompl/base/PlannerTerminationCondition.h>
#include <ompl/geometric/planners/rrt/RRTConnect.h>
#include <ompl/geometric/planners/rrt/RRTstar.h>
#include <ompl/geometric/planners/fmt/FMT.h>
#include <ompl/geometric/planners/prm/PRM.h>
#include <ompl/geometric/planners/rrt/RRTConnect.h>
#include <ompl/geometric/planners/rrt/RRT.h>
#include <ompl/geometric/planners/kpiece/KPIECE1.h>
#include <ompl/geometric/planners/kpiece/LBKPIECE1.h>
#include <ompl/vamp/VampStateValidityChecker.h>
#include <ompl/vamp/VampMotionValidator.h>
#include <ompl/geometric/SimpleSetup.h>
#include <ompl/vamp/VampStateSpace.h>
#include <vamp/robots/panda.hh>
#include <vamp/vector.hh>
#include <ompl/tools/benchmark/Benchmark.h>


#include <iostream>
#include <iomanip>
#include <numeric>
#include <algorithm>
#include <cmath>
#include <fstream>

namespace ob = ompl::base;
namespace og = ompl::geometric;
namespace ot = ompl::tools;


MotionBenchmakerDemo::MotionBenchmakerDemo(const std::string& robotName,
                                          const std::string& problemFile,
                                          const std::string& plannerName)
    : robotName_(robotName), plannerName_(plannerName)
{
    loadProblemsFromJSON(problemFile);
    initializeStateSpace();
}

void MotionBenchmakerDemo::loadProblemsFromJSON(const std::string& filename)
{
    using boost::property_tree::ptree;
    ptree pt;
    
    try {
        std::ifstream file(filename);
        if (!file.is_open()) {
            throw std::runtime_error("Failed to open JSON file: " + filename);
        }
        boost::property_tree::json_parser::read_json(file, pt);
    } catch (const std::exception& e) {
        throw std::runtime_error(std::string("Failed to load JSON file: ") + e.what());
    }

    // Check if problems key exists
    auto problemsIt = pt.find("problems");
    if (problemsIt == pt.not_found()) {
        throw std::runtime_error("JSON file does not contain 'problems' key");
    }

    // Iterate through each problem type (e.g., "problem_type_1", "problem_type_2")
    for (const auto& problemEntry : pt.get_child("problems")) {
        std::string problemName = problemEntry.first;
        problemNames_.push_back(problemName);

        const auto& problemInstances = problemEntry.second;
        std::vector<ptree> instances;

        // Each problem type contains an array of problem instances
        for (const auto& instanceEntry : problemInstances) {
            instances.push_back(instanceEntry.second);
        }
        
        problems_[problemName] = instances;
    }

    std::cout << "Loaded " << problemNames_.size() << " problem types from " << filename << std::endl;
}

void MotionBenchmakerDemo::initializeStateSpace()
{
    // Create state space
    using Robot = vamp::robots::Panda;
    space_ = std::make_shared<ompl::vamp::VampStateSpace<Robot>>();
}

std::shared_ptr<ompl::base::Planner> MotionBenchmakerDemo::createPlanner(
    const std::shared_ptr<ompl::base::SpaceInformation>& si) const
{
    std::shared_ptr<ompl::base::Planner> planner;

    if (plannerName_ == "RRTConnect" || plannerName_ == "rrtc") {
        planner = std::make_shared<og::RRTConnect>(si);
    } else if (plannerName_ == "RRTstar" || plannerName_ == "rrtstar") {
        planner = std::make_shared<og::RRTstar>(si);
    } else if (plannerName_ == "FMT" || plannerName_ == "fmt") {
        planner = std::make_shared<og::FMT>(si);
    } else if (plannerName_ == "PRM" || plannerName_ == "prm") {
        planner = std::make_shared<og::PRM>(si);
    }
     else if (plannerName_ == "KPIECE1" || plannerName_ == "kpiece1") {
        planner = std::make_shared<og::KPIECE1>(si);
    }
    else if (plannerName_ == "RRT" || plannerName_ == "rrt") {
        planner = std::make_shared<og::RRT>(si);
    } else if (plannerName_ == "LBKPIECE1" || plannerName_ == "lbkpiece1") {
        planner = std::make_shared<og::LBKPIECE1>(si);
    }
    else {
        std::cerr << "Unknown planner '" << plannerName_ << "', using RRTConnect" << std::endl;
        planner = std::make_shared<og::RRTConnect>(si);
    }

    return planner;
}

bool MotionBenchmakerDemo::setupProblem(
    const std::string& problemName,
    const boost::property_tree::ptree& problemData)
{

    // Check if problem is valid
    bool valid = problemData.get<bool>("valid", true);
    if (!valid) {
        return false;
    }

    // Build environment from problem data
    vamp::collision::Environment<float> env_float;
    bool is_box = problemName == "box" ? true : false;
    
    // Load spheres
    if (problemData.find("sphere") != problemData.not_found()) {
        for (const auto& sphereEntry : problemData.get_child("sphere")) {
            const auto& s = sphereEntry.second;
            std::array<float, 3> center;
            size_t i = 0;
            const auto& posChild = (s.find("center") != s.not_found()) ? 
                s.get_child("center") : s.get_child("position");
            for (const auto& coord : posChild) {
                if (i < 3) center[i++] = std::stof(coord.second.data());
            }
            float radius = s.get<float>("radius");
            env_float.spheres.emplace_back(
                vamp::collision::factory::sphere::array(center, radius));
        }
    }
    
    // Load cylinders (convert to capsules)
    if (problemData.find("cylinder") != problemData.not_found()) {
        for (const auto& cylEntry : problemData.get_child("cylinder")) {
            const auto& c = cylEntry.second;
            std::array<float, 3> center, orientation;
            size_t i = 0;
            const auto& posChild = (c.find("center") != c.not_found()) ? 
                c.get_child("center") : c.get_child("position");
            for (const auto& coord : posChild) {
                if (i < 3) center[i++] = std::stof(coord.second.data());
            }
            i = 0;
            for (const auto& euler : c.get_child("orientation_euler_xyz")) {
                if (i < 3) orientation[i++] = std::stof(euler.second.data());
            }
            float radius = c.get<float>("radius");
            float length = c.get<float>("length");
            if (is_box) {
                std::array<float, 3> halfExtents = {radius, radius, length / 2.0};
                env_float.cuboids.emplace_back(
                    vamp::collision::factory::cuboid::array(center, orientation, halfExtents));
            } 
            else {
                env_float.capsules.emplace_back(
                    vamp::collision::factory::capsule::center::array(center, orientation, radius, length));
            }
        }
    }
    
    // Load boxes
    if (problemData.find("box") != problemData.not_found()) {
        for (const auto& boxEntry : problemData.get_child("box")) {
            const auto& b = boxEntry.second;
            std::array<float, 3> center, orientation, halfExtents;
            size_t i = 0;
            const auto& posChild = (b.find("center") != b.not_found()) ? 
                b.get_child("center") : b.get_child("position");
            for (const auto& coord : posChild) {
                if (i < 3) center[i++] = std::stof(coord.second.data());
            }
            i = 0;
            for (const auto& euler : b.get_child("orientation_euler_xyz")) {
                if (i < 3) orientation[i++] = std::stof(euler.second.data());
            }
            i = 0;
            for (const auto& extent : b.get_child("half_extents")) {
                if (i < 3) halfExtents[i++] = std::stof(extent.second.data());
            }
            env_float.cuboids.emplace_back(
                vamp::collision::factory::cuboid::array(center, orientation, halfExtents));
        }
    }

    using Environment = vamp::collision::Environment<vamp::FloatVector<vamp::FloatVectorWidth>>;
    currentEnv_ = std::make_shared<Environment>(env_float);

    using Robot = vamp::robots::Panda;
    auto space_info = ss_->getSpaceInformation();

    // Add VAMP-based validators using the environment
    space_info->setStateValidityChecker(
        std::make_shared<ompl::vamp::VampStateValidityChecker<Robot>>(space_info, *currentEnv_));
    space_info->setMotionValidator(
        std::make_shared<ompl::vamp::VampMotionValidator<Robot>>(space_info, *currentEnv_));

    // Set start and goal configurations
    ob::ScopedState<> start(space_);
    ob::ScopedState<> goal(space_);

    // Get start configuration
    std::vector<double> startVec;
    if (problemData.find("start") != problemData.not_found()) {
        for (const auto& val : problemData.get_child("start")) {
            startVec.push_back(std::stod(val.second.data()));
        }
    }
    
    for (size_t i = 0; i < startVec.size() && i < space_->getDimension(); ++i) {
        start[i] = startVec[i];
    }

    // Get goal configuration
    std::vector<double> goalVec;
    if (problemData.find("goal") != problemData.not_found()) {
        for (const auto& val : problemData.get_child("goal")) {
            goalVec.push_back(std::stod(val.second.data()));
        }
    } else if (problemData.find("goals") != problemData.not_found()) {
        // Use first goal from goals array
        auto goalsChild = problemData.get_child("goals");
        if (goalsChild.size() > 0) {
            auto firstGoal = goalsChild.begin()->second;
            for (const auto& val : firstGoal) {
                goalVec.push_back(std::stod(val.second.data()));
            }
        }
    }
    
    for (size_t i = 0; i < goalVec.size() && i < space_->getDimension(); ++i) {
        goal[i] = goalVec[i];
    }

    // print start vs goal
    // std::cout << "Start: ";
    // for (size_t i = 0; i < startVec.size(); ++i) {
    //     std::cout << start[i] << " ";
    // }
    // std::cout << std::endl;

    // std::cout << "Goal: ";
    // for (size_t i = 0; i < goalVec.size(); ++i) {
    //     std::cout << goal[i] << " ";
    // }
    // std::cout << std::endl;

    // Set the start and goal states
    ss_->setStartAndGoalStates(start, goal);
    return true;
}


PlanningResult MotionBenchmakerDemo::solveInstance(
    const std::string& problemName,
    const boost::property_tree::ptree& problemData,
    double timeoutSeconds)
{
    PlanningResult result;
    ss_ = std::make_shared<og::SimpleSetup>(space_);
    
    auto space_info = ss_->getSpaceInformation();
    
    if (!setupProblem(problemName, problemData)) {
        return result;
    }
    space_info = ss_->getSpaceInformation();
    // Setup benchmarking
    auto planner = createPlanner(space_info);
    ss_->setPlanner(planner);

    // Plan
    auto startTime = std::chrono::steady_clock::now();
    ob::PlannerStatus status = ss_->solve(
        ob::timedPlannerTerminationCondition(timeoutSeconds)
    );
    auto endTime = std::chrono::steady_clock::now();

    result.planningTime = std::chrono::duration_cast<std::chrono::nanoseconds>(
        endTime - startTime
    );

    if (status == ob::PlannerStatus::EXACT_SOLUTION ||
        status == ob::PlannerStatus::APPROXIMATE_SOLUTION) {
        result.solved = true;
        auto path = ss_->getSolutionPath();
        result.pathVertices = path.getStateCount();
        result.pathCost = path.length();
    }

    // Get planner data for iteration count
    ob::PlannerData plannerData(space_info);
    ss_->getPlanner()->getPlannerData(plannerData);
    result.planningIterations = plannerData.numVertices();

    // Clean up
    ss_.reset();
    
    return result;
}

PlanningResult MotionBenchmakerDemo::benchmarkInstance(
    const std::string& problemName,
    const boost::property_tree::ptree& problemData,
    unsigned int benchmarkTrials,
    double timeoutSeconds)
{
    PlanningResult result;
    ss_ = std::make_shared<og::SimpleSetup>(space_);
    
    if (!setupProblem(problemName, problemData)) {
        return result;
    }
    auto space_info = ss_->getSpaceInformation();
    // setup benchmark
    double memoryLimit = 4096;

    std::string benchmarkName = "MBM-" + problemName;
    ot::Benchmark b(*ss_, benchmarkName);
    ot::Benchmark::Request request(timeoutSeconds, memoryLimit, benchmarkTrials);

    b.addPlanner(std::make_shared<og::RRTConnect>(space_info));
    b.addPlanner(std::make_shared<og::RRT>(space_info));
    b.addPlanner(std::make_shared<og::KPIECE1>(space_info));    
    b.addPlanner(std::make_shared<og::LBKPIECE1>(space_info));

    b.benchmark(request);
    b.saveResultsToFile("vamp_mbm_benchmark.log");
    result.solved = true;
    ss_.reset();
    
    return result;
}


std::vector<PlanningResult> MotionBenchmakerDemo::benchmarkProblem(
    const std::string& problemName,
    unsigned int benchmarkTrials,
    double timeoutSeconds)
{
    std::vector<PlanningResult> allResults;

    if (problems_.find(problemName) == problems_.end()) {
        std::cerr << "Problem '" << problemName << "' not found" << std::endl;
        return allResults;
    }

    const auto& instances = problems_[problemName];
    std::cout << "Benchmarking '" << problemName << "' with " << instances.size()
              << " instances..." << std::endl;

    for (size_t i = 0; i < instances.size(); ++i) {
        std::cout << "  Instance " << i + 1 << "/" << instances.size() << "... ";
        std::cout.flush();

        if (benchmarkTrials == 0) {
            auto result = solveInstance(problemName, instances[i], timeoutSeconds);
            allResults.push_back(result);
        } 
        else {
            auto result = benchmarkInstance(problemName, instances[i], benchmarkTrials, timeoutSeconds);
            allResults.push_back(result);
            break;
        }
    }

    return allResults;
}

std::map<std::string, std::vector<PlanningResult>> MotionBenchmakerDemo::benchmarkAll(
    unsigned int benchmarkTrials,
    double timeoutSeconds,
    bool print_failures)
{
    std::map<std::string, std::vector<PlanningResult>> allResults;

    for (const auto& problemName : problemNames_) {

        // if (problemName != "table_under_pick") {
        //     continue;
        // }

        auto results = benchmarkProblem(problemName, benchmarkTrials, timeoutSeconds);
        allResults[problemName] = results;

        if (print_failures) {
            unsigned int failures = 0;
            for (const auto& result : results) {
                if (!result.solved) {
                    failures++;
                }
            }
            if (failures > 0) {
                std::cout << "    Failed: " << failures << "/" << results.size() << std::endl;
            }
        }

    }

    return allResults;
}

void MotionBenchmakerDemo::printStatistics(const std::string& problemName,
                                          const std::vector<PlanningResult>& results)
{
    if (results.empty()) {
        std::cout << "No results for problem '" << problemName << "'" << std::endl;
        return;
    }

    // Count successes
    unsigned int successes = 0;
    std::vector<double> planningTimes;
    std::vector<double> pathCosts;
    std::vector<unsigned int> iterations;
    std::vector<unsigned int> vertices;

    for (const auto& result : results) {
        if (result.solved) {
            successes++;
            planningTimes.push_back(result.planningTime.count() / 1e9);  // Convert to seconds
            pathCosts.push_back(result.pathCost);
            iterations.push_back(result.planningIterations);
            vertices.push_back(result.pathVertices);
        }
    }

    std::cout << "\n=== Statistics for '" << problemName << "' ===" << std::endl;
    std::cout << "Success Rate: " << successes << "/" << results.size() << std::endl;

    if (successes > 0) {
        // Calculate statistics
        auto minMaxTime = std::minmax_element(planningTimes.begin(), planningTimes.end());
        double meanTime = std::accumulate(planningTimes.begin(), planningTimes.end(), 0.0) /
                         planningTimes.size();

        auto minMaxCost = std::minmax_element(pathCosts.begin(), pathCosts.end());
        double meanCost = std::accumulate(pathCosts.begin(), pathCosts.end(), 0.0) /
                         pathCosts.size();

        std::cout << std::fixed << std::setprecision(4);
        std::cout << "Planning Time (s): min=" << *minMaxTime.first
                  << ", max=" << *minMaxTime.second
                  << ", mean=" << meanTime << std::endl;

        std::cout << "Path Cost (m): min=" << *minMaxCost.first
                  << ", max=" << *minMaxCost.second
                  << ", mean=" << meanCost << std::endl;

        unsigned int meanIter = std::accumulate(iterations.begin(), iterations.end(), 0u) /
                               iterations.size();
        unsigned int meanVert = std::accumulate(vertices.begin(), vertices.end(), 0u) /
                               vertices.size();

        std::cout << "Mean Iterations: " << meanIter << std::endl;
        std::cout << "Mean Path Vertices: " << meanVert << std::endl;
    }
    std::cout << std::endl;
}

File: demos/vamp/motion_benchmaker_demo.py
import pickle
import time
from tabulate import tabulate
from tqdm import tqdm
from pathlib import Path
import pandas as pd
from typing import Union, List
from functools import partial
import sys
from datetime import datetime
import subprocess
import numpy as np

from fire import Fire
import vamp
from vamp import pointcloud as vpc
from ompl import base as ob
from ompl import geometric as og
from ompl import tools as ot

from vamp_state_space import VampStateSpace, VampMotionValidator, VampStateValidityChecker

sys.path.insert(0, str(Path(__file__).parent.parent))
from viser_visualizer import ViserVisualizer

def main(
    robot: str = "panda",                  # Robot to plan for
    planner: str = "rrtc",                 # Planner name to use (e.g., 'rrtc', 'prm', 'rrt')
    dataset: str = "problems.pkl",         # Pickled dataset to use
    problem: Union[str, List[str]] = [],   # Problem name or list of problems to evaluate
    trials: int = 1,                       # Number of trials to evaluate each instance
    print_failures: bool = False,          # Print out failures and invalid problems
    visualize: bool = False,              # Visualize solutions using Viser (any key=next, q=disable, w=skip problem set)
    pointcloud: bool = False,              # Use pointcloud rather than primitive geometry
    samples_per_object: int = 10000,       # If pointcloud, samples per object to use
    filter_radius: float = 0.02,           # Filter radius for pointcloud filtering
    filter_cull: bool = True,              # Cull pointcloud around robot by maximum distance
    planning_time: float = 1.0,            # Planning time limit in seconds
    benchmark: int = 0,                   # Run OMPL benchmark for n trials instead of 700 problems
    n_trials: int = 100,                   # Number of trials for benchmarking
    **kwargs,
    ):

    if robot not in vamp.robots:
        raise RuntimeError(f"Robot {robot} does not exist in VAMP!")

    # Get robot module
    robot_module = getattr(vamp, robot)
    dimension = robot_module.dimension()
    
    # Setup OMPL state space
    space = ob.RealVectorStateSpace(dimension)
    bounds = ob.RealVectorBounds(dimension)
    upper_bounds = robot_module.upper_bounds()
    lower_bounds = robot_module.lower_bounds()
    for i in range(dimension):
        bounds.setLow(i, lower_bounds[i])
        bounds.setHigh(i, upper_bounds[i])
    space.setBounds(bounds)
    
    # Map planner names to OMPL planners
    planner_map = {
        'rrtc': og.RRTConnect,
        'rrt': og.RRT,
        'prm': og.PRM,
        'rrtstar': og.RRTstar,
        'bitstar': og.BITstar,
    }
    
    if planner.lower() not in planner_map:
        raise RuntimeError(f"Planner {planner} not recognized. Available: {list(planner_map.keys())}")
    
    planner_class = planner_map[planner.lower()]
    vamp_folder = Path(__file__).parent.parent.parent / 'external' / 'vamp'
    problems_dir = vamp_folder / 'resources' / robot
    pickle_path = problems_dir / dataset
    
    # Check if pickle file exists, generate if not
    if not pickle_path.exists():
        print(f"Pickle file not found at {pickle_path}, generating from tar.bz2...")
        script_path = vamp_folder / 'resources' / 'problem_tar_to_pkl_json.py'
        result = subprocess.run([sys.executable, str(script_path), f'--robot={robot}'], check=True)
        if result.returncode != 0:
            raise RuntimeError(f"Failed to generate pickle file using {script_path}")
        print(f"Successfully generated pickle file")
    
    with open(pickle_path, 'rb') as f:
        problems = pickle.load(f)

    problem_names = list(problems['problems'].keys())
    if isinstance(problem, str):
        problem = [problem]

    if not problem:
        problem = problem_names
    else:
        for problem_name in problem:
            if problem_name not in problem_names:
                raise RuntimeError(
                    f"Problem `{problem_name}` not available! Available problems: {problem_names}"
                    )

    total_problems = 0
    valid_problems = 0
    failed_problems = 0
    
    # Initialize visualizer if requested
    vis = None
    visualize_enabled = visualize
    skip_to_next_problem_set = False
    if visualize:
        vis = ViserVisualizer(robot_name=robot, robot_dimension=dimension)
        # vis.add_grid()
    
    if benchmark:    
        # Create benchmark output directory with timestamp
        timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
        benchmark_dir = Path("benchmarks") / f"benchmark_{timestamp}"
        benchmark_dir.mkdir(parents=True, exist_ok=True)
        
        print(f"Benchmarking {robot} on first problem of each type with {n_trials} trials...")
        print(f"Results will be saved to: {benchmark_dir}")
        
        # Set this to x plans to do
        planner_constructors = [
            og.RRTConnect,
            og.PRM,
            og.KPIECE1,
            og.LBKPIECE1,
        ]
        
    tick = time.perf_counter()
    results = []
    
    print(f"Evaluating {planner_class.__name__} on problems: {problems['problems'].keys()} with robot {robot}...")
    
    for name, pset in problems['problems'].items():
        if name not in problem:
            continue
        
        # Reset skip flag for new problem set
        skip_to_next_problem_set = False

        failures = []
        invalids = []
        print(f"Evaluating {robot} on {name}: ")
        progress_bar = tqdm(total=len(pset), desc=f"Processing {name}")
        for i, data in enumerate(pset):
            total_problems += 1
            skip_pset = False
            progress_bar.update(1)

            if not data['valid']:
                invalids.append(i)
                continue

            valid_problems += 1

            if pointcloud:
                r_min, r_max = robot_module.min_max_radii()
                (env, original_pc, filtered_pc, filter_time, build_time) = vpc.problem_dict_to_pointcloud(
                    robot,
                    r_min,
                    r_max,
                    data,
                    samples_per_object,
                    filter_radius,
                    filter_cull,
                    )

                pointcloud_results = {
                    'original_pointcloud_size': len(original_pc),
                    'filtered_pointcloud_size': len(filtered_pc),
                    'filter_time': pd.Timedelta(nanoseconds = filter_time),
                    'capt_build_time': pd.Timedelta(nanoseconds = build_time)
                    }
            else:
                env = vamp.problem_dict_to_vamp(data)

            for trial in range(trials):
                # Setup OMPL problem
                dimension = robot_module.dimension()
                space = VampStateSpace(robot=robot_module)
                si = ob.SpaceInformation(space)
                
                motion_validator = VampMotionValidator(si = si, env = env, robot = robot_module)
                state_validity_checker = VampStateValidityChecker(si = si, env = env, robot = robot_module)
                
                # Set validators
                si.setMotionValidator(motion_validator)
                si.setStateValidityChecker(state_validity_checker)
                
                ss = og.SimpleSetup(si)
                ompl_planner = planner_class(si)
                ss.setPlanner(ompl_planner)
                
                # Set start state
                start = si.allocState()
                start[0:dimension] = data['start']
                
                # Set goal state (use first goal)
                goal = si.allocState()
                goal[0:dimension] = data['goals'][0]
                
                ss.setStartAndGoalStates(start, goal)
                
                # Benchmark
                if benchmark:
                    benchmark_name = f"{robot}_{name}"
                    ompl_benchmark = ot.Benchmark(ss, benchmark_name)
            
                    for planner_constructor in planner_constructors:
                        ompl_benchmark.addPlanner(planner_constructor(si))
                    
                    # Configure benchmark request
                    req = ot.Request()
                    req.maxTime = planning_time
                    req.maxMem = 1000.0
                    req.runCount = n_trials
                    req.displayProgress = True
                    
                    print(f"  Running benchmark with {n_trials} trials...")
                    ompl_benchmark.benchmark(req)
                    
                    # Save results
                    file = str(f"vamp_mbm_python")
                    log_file = file + ".log"
                    db_file = file + ".db"
                    
                    ompl_benchmark.saveResultsToFile(log_file)
                    print(f"  Saved log to: {log_file}")
                    
                    # Convert log to database
                    ot.readBenchmarkLog(db_file, [log_file], moveitformat=False)
                    print(f"  Saved database to: {db_file}")
                    # only run one per problem
                    skip_pset = True
                    break
                # Single Plan
                planning_start = time.perf_counter()
                result = ss.solve(planning_time)
                planning_elapsed = time.perf_counter() - planning_start
                
                if not ss.haveExactSolutionPath():
                    failures.append(i)
                    print(f"Failed to find exact solution for problem {i}")
                    break

                # Get path
                path = ss.getSolutionPath()
                initial_cost = path.length()
                
                # Simplify
                simplify_start = time.perf_counter()
                ss.simplifySolution()
                simplify_elapsed = time.perf_counter() - simplify_start
                
                simplified_path = ss.getSolutionPath()
                simplified_cost = simplified_path.length()
                
                trial_result = {
                    'planning_time': pd.Timedelta(microseconds=planning_elapsed * 1e6),
                    'simplification_time': pd.Timedelta(microseconds=simplify_elapsed * 1e6),
                    'total_time': pd.Timedelta(microseconds=(planning_elapsed + simplify_elapsed) * 1e6),
                    'initial_path_cost': initial_cost,
                    'simplified_path_cost': simplified_cost,
                    'planning_iterations': 0,  # OMPL doesn't directly expose this
                    'avg_time_per_iteration': 0.0,
                }
                
                if pointcloud:
                    trial_result.update(pointcloud_results)

                results.append(trial_result)
                
                # Visualization
                if visualize_enabled and vis is not None:
                    # Check if we should skip to next problem set
                    if skip_to_next_problem_set:
                        continue
                    # Clear previous visualization
                    vis.reset()
                    # vis.add_grid()
                    
                    # Load environment obstacles
                    if pointcloud:
                        original_pc = np.array(original_pc)
                        z_values = original_pc[:, 2]
                        z_min, z_max = z_values.min(), z_values.max()
                        z_range = z_max - z_min
                        
                        # Define color range: start at 20% and end at 80% of z-range
                        color_start = z_min + 0.2 * z_range
                        color_end = z_min + 1.0 * z_range
                        
                        # Clamp and normalize z values to [0, 1] within the color range
                        z_normalized = np.clip((z_values - color_start) / (color_end - color_start), 0, 1)
                        
                        # Smooth rainbow colormap: red -> yellow -> green -> cyan -> blue
                        colors = np.zeros((len(original_pc), 3))
                        colors[:, 0] = np.maximum(0, 1 - 2 * np.abs(z_normalized - 0.25))  # Red
                        colors[:, 1] = np.maximum(0, 1 - 2 * np.abs(z_normalized - 0.5))   # Green
                        colors[:, 2] = np.maximum(0, 1 - 2 * np.abs(z_normalized - 0.75))  # Blue
                        vis.add_point_cloud(original_pc, color=colors, point_size=0.01)
                    else:
                        vis.load_mbm_environment(data, padding=0.0, color=(0.8, 0.4, 0.2, 0.75))
                    
                    # Convert path to numpy array
                    simplified_path.interpolate(150)
                    states = simplified_path.getStates()
                    trajectory = np.array([list(state[0:dimension]) for state in states])
                    # interpolate
                    
                    
                    vis.visualize_trajectory(trajectory)
                    
                    print(f"\nVisualizing problem {name} [{i}] - Press any key to continue, 'q' to disable viz, 'w' to skip to next problem set")
                    key = vis.play_until_key_pressed(key='any', dt=1/60)
                    
                    if key == 'q':
                        print("Visualization disabled for remaining problems")
                        visualize_enabled = False
                    elif key == 'w':
                        print(f"Skipping to next problem set...")
                        skip_to_next_problem_set = True
                        break 
            
            if skip_pset:
                break
        
        progress_bar.close()
        
        failed_problems += len(failures)

        if print_failures:
            if invalids:
                print(f"  Invalid problems: {invalids}")

            if failures:
                print(f"  Failed on {failures}")
                
    
    if benchmark > 0:
        print(f"All benchmarks saved in {benchmark_dir}")
        exit(0)
    
    tock = time.perf_counter()

    df = pd.DataFrame.from_dict(results)

    # Convert to microseconds
    df["planning_time"] = df["planning_time"].dt.microseconds
    df["simplification_time"] = df["simplification_time"].dt.microseconds
    df["avg_time_per_iteration"] = df["planning_iterations"] / df["planning_time"]

    # Pointcloud data
    if pointcloud:
        df["total_build_and_plan_time"] = df["total_time"] + df["filter_time"] + df["capt_build_time"]
        df["filter_time"] = df["filter_time"].dt.microseconds / 1e3
        df["capt_build_time"] = df["capt_build_time"].dt.microseconds / 1e3
        df["total_build_and_plan_time"] = df["total_build_and_plan_time"].dt.microseconds / 1e3

    df["total_time"] = df["total_time"].dt.microseconds

    # Get summary statistics
    time_stats = df[[
        "planning_time",
        "simplification_time",
        "total_time",
        "planning_iterations",
        "avg_time_per_iteration",
        ]].describe(percentiles = [0.25, 0.5, 0.75, 0.95])
    time_stats.drop(index = ["count"], inplace = True)

    cost_stats = df[[
        "initial_path_cost",
        "simplified_path_cost",
        ]].describe(percentiles = [0.25, 0.5, 0.75, 0.95])
    cost_stats.drop(index = ["count"], inplace = True)

    if pointcloud:
        pointcloud_stats = df[[
            "filter_time",
            "capt_build_time",
            "total_build_and_plan_time",
            ]].describe(percentiles = [0.25, 0.5, 0.75, 0.95])
        pointcloud_stats.drop(index = ["count"], inplace = True)

    print()
    print(
        tabulate(
            time_stats,
            headers = [
                'Planning Time (μs)',
                'Simplification Time (μs)',
                'Total Time (μs)',
                'Planning Iters.',
                'Time per Iter. (μs)',
                ],
            tablefmt = 'github'
            )
        )

    print(
        tabulate(
            cost_stats, headers = [
                ' Initial Cost (L2)',
                '    Simplified Cost (L2)',
                ], tablefmt = 'github'
            )
        )

    if pointcloud:
        print(
            tabulate(
                pointcloud_stats,
                headers = [
                    '  Filter Time (ms)',
                    '    CAPT Build Time (ms)',
                    'Total Time (ms)',
                    ],
                tablefmt = 'github'
                )
            )

    print(
        f"Solved / Valid / Total # Problems: {valid_problems - failed_problems} / {valid_problems} / {total_problems}"
        )
    print(f"Completed all problems in {df['total_time'].sum() / 1000:.3f} milliseconds")
    print(f"Total time including Python overhead: {(tock - tick) * 1000:.3f} milliseconds")


if __name__ == "__main__":
    Fire(main)
